File: 백준/Gold/2042. 구간 합 구하기/구간 합 구하기.java
import java.io.*;
import java.util.*;

public class Main {
    static long[] tree;
    public static void main(String[] args) throws IOException {
        try (BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
             BufferedWriter bw = new BufferedWriter(new OutputStreamWriter(System.out))) {
            StringTokenizer st = new StringTokenizer(br.readLine());
            int N = Integer.parseInt(st.nextToken()),
                    M = Integer.parseInt(st.nextToken()),
                    K = Integer.parseInt(st.nextToken());
            long[] array = new long[N];
            for (int i = 0; i < N; i++) array[i] = Long.parseLong(br.readLine());
            int h = (int) Math.ceil(Math.log(N) / Math.log(2));
            tree = new long[(int) Math.pow(2, h + 1)];
            generate(array, 1, 0, N - 1);
            StringBuilder print = new StringBuilder();
            while (M + K > 0) {
                st = new StringTokenizer(br.readLine());
                int a = Integer.parseInt(st.nextToken()), b = Integer.parseInt(st.nextToken());
                if (a == 1) {
                    long c = Long.parseLong(st.nextToken());
                    update(1, 0, N - 1, b - 1, c - array[b - 1]);
                    array[b - 1] = c;
                    M--;
                } else {
                    int c = Integer.parseInt(st.nextToken());
                    print.append(prefixSum(1, 0, N - 1, b - 1, c - 1)).append('\n');
                    K--;
                }
            }
            bw.write(print.toString());
            bw.flush();
        }
    }
    public static long generate(long[] arr, int current, int start, int end) {
        if (start == end) return tree[current] = arr[start];
        return tree[current] = generate(arr, current * 2, start, (start + end) / 2)
                + generate(arr, current * 2 + 1, (start + end) / 2 + 1, end);
    }
    public static long prefixSum(int current, int start, int end, int left, int right) {
        if (left > end || right < start) return 0;
        if (left <= start && right >= end) return tree[current];
        return prefixSum(current * 2, start, (start + end) / 2, left, right)
                + prefixSum(current * 2 + 1, (start + end) / 2 + 1, end, left, right);
    }
    public static void update(int current, int start, int end, int idx, long diff) {
        if (idx < start || idx > end) return;
        tree[current] += diff;
        if (start != end) {
            update(current * 2, start, (start + end) / 2, idx, diff);
            update(current * 2 + 1, (start + end) / 2 + 1, end, idx, diff);
        }
    }
}

File: 백준/Gold/2042. 구간 합 구하기/구간 합 구하기.cc
#include <iostream>
#include <vector>
#include <cmath>

using namespace std;

#define ll long long

class SegmentTree {
    vector<ll> tree;
    vector<ll> array;
    int arrLength;
    ll generate(int current, int start, int end, vector<ll>& arr);
    ll query(int current, int start, int end, int left, int right);
    void update(int current, int start, int end, int idx, ll value);
public:
    SegmentTree(vector<ll>& arr);
    ll query(int left, int right);
    void update(int idx, ll value);
};

SegmentTree::SegmentTree(vector<long long>& arr) {
    array = arr;
    arrLength = arr.size();
    int h = (int) ceil(log2(arrLength));
    tree.resize(2 << h);
    generate(1, 0, arrLength - 1, arr);
}

ll SegmentTree::generate(int current, int start, int end, vector<long long>& arr) {
    if (start == end) return this->tree[current] = arr[start];;
    int mid = (start + end) >> 1;
    return this->tree[current] =
        generate(current * 2, start, mid, arr) +
        generate(current * 2 + 1, mid + 1, end, arr);
}

ll SegmentTree::query(int current, int start, int end, int left, int right) {
    if (left > end || right < start) return 0;
    if (left <= start && right >= end) return this->tree[current];
    int mid = (start + end) >> 1;
    return query(current * 2, start, mid, left, right) +
        query(current * 2 + 1, mid + 1, end, left, right);
}

void SegmentTree::update(int current, int start, int end, int idx, long long value) {
    if (idx < start || idx > end) return;
    tree[current] += value;
    if (start != end) {
        int mid = (start + end) >> 1;
        update(current * 2, start, mid, idx, value);
        update(current * 2 + 1, mid + 1, end, idx, value);
    }
}

ll SegmentTree::query(int left, int right) {
    return query(1, 0, arrLength - 1, left - 1, right - 1);
}

void SegmentTree::update(int idx, long long value) {
    update(1, 0, arrLength - 1, idx - 1, value - array[idx - 1]);
    array[idx - 1] = value;
}

int main() {
    ios::sync_with_stdio(false);
    cin.tie(NULL);

    int n, m, k;
    cin >> n >> m >> k;

    vector<ll> arr(n);
    for (int i = 0; i < n; ++i) {
        cin >> arr[i];
    }

    SegmentTree obj(arr);

    ll q, a, b;
    while (m + k > 0) {
        cin >> q >> a >> b;
        if (q == 1) {
            obj.update(a, b);
            m--;
        } else {
            cout << obj.query(a, b) << '\n';
            k--;
        }
    }

    return 0;
}